File: Server/BL/DonorService.cs
using Microsoft.AspNetCore.Mvc;
using Server.BL.Interfaces;
using Server.DAL;
using Server.DAL.Interfaces;
using Server.Models;
using Server.Models.DTO;

namespace Server.BL
{
    public class DonorService:IDonorService
    {
        public readonly IDonorDal donorDal;
        public DonorService(IDonorDal donorDal)
        {
            this.donorDal = donorDal;
        }

        public async Task AddDonor(DonorDTO value)
        {
            if(value != null)
            {
                //if (value.Name.Trim().Length <= 50 && value.Name.Trim().Length >= 2)
                //     if(value.phone.Trim().Length >= 7 && value.phone.Trim().Length > 15)
                //    {

                //    }

            }             
             await donorDal.AddDonor(value);
        }

        public async Task DeleteDonor(int Id)

        {
          await donorDal.DeleteDonor(Id);
        }

        public async Task<List<Donor>> FilterDonors(string? name, string? present)
        {
            return await donorDal.FilterDonors(name, present);
        }

        public async Task<List<Donor>> GetAll()
        {
            return await donorDal.GetAll();
        }

        public async Task<Donor>  GetById(int Id)
        {
            return await donorDal.GetById(Id);
        }

        public async Task<List<Present>> GetPresentsDonor(int donorId)
        {
            return await donorDal.GetPresentsDonor(donorId);
        }

        public async Task UpdateDonor(DonorDTO value, int Id)
        {
             await donorDal.UpdateDonor(value, Id);
        }

    }
}

    

File: client/src/app/services/donor.service.ts

import { Injectable, inject } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';

import { Observable, throwError } from 'rxjs';

import { Donor } from '../models/DonorModel';
import { DonorDTO } from '../models/DTO/DonorDTO';
import { Present } from '../models/PresentModel';

@Injectable({
  providedIn: 'root'
})
export class DonorService {

  BASE_URL = 'http://localhost:5000/api/Donor';
  httpClient: HttpClient = inject(HttpClient);
  constructor() { }
  GetAllDonors(): Observable<Donor[]> {
    var res = this.httpClient.get<Donor[]>(this.BASE_URL);
    return res;
  }
  
  deleteDonor(id: number): Observable<void> {

    return this.httpClient.delete<void>(`${this.BASE_URL}/${id}`)
  }

  UpdateDonor(p: DonorDTO, id: number): Observable<void> {

    return this.httpClient.put<void>(`${this.BASE_URL}/${id}`, p);

  }

  AddDonor(value: DonorDTO): Observable<void> {
    return this.httpClient.post<void>(this.BASE_URL, value);

  }
  GetPresentsDonor(donorId:number): Observable<Present[]> {
    var res = this.httpClient.get<Present[]>(`${this.BASE_URL}/presentsDonor/${donorId}`);
    return res;
  }
//   FilterDonors(kind:string,value:string): Observable<Donor[]> {
// console.log(value);

//     var res = this.httpClient.get<Donor[]>(`${this.BASE_URL}/FilterDonors/${value}`);
//     console.log(res);
    
//     return res;
//   }
FilterDonors(name: string | '', present: string | ''): Observable<Donor[]> {
  // הכנת פרמטרים לשאילתה
  const params = new HttpParams()
    .set('name', name ?? '') // אם name הוא null, ישלח כערך ריק
    .set('present', present ?? ''); // כנ"ל עבור present

  // שליחת בקשה לשרת עם פרמטרים
  return this.httpClient.get<Donor[]>(`${this.BASE_URL}/FilterDonors`, { params });
}

}
